src/analysis_and_optimization/Mir_utils.ml

```diff
@@ -298,3 +298,29 @@ let rec update_expr_ad_levels autodiffable_variables e =
 
 and update_idx_ad_levels autodiffable_variables =
   map_index (update_expr_ad_levels autodiffable_variables)
+
+let number_locations_prog (mir : typed_prog) :
+    typed_prog_num * string Int.Table.t =
+  let module LocSp = struct
+    type t = location_span
+
+    let compare = compare_location_span
+    let hash = hash_location_span
+    let sexp_of_t = sexp_of_location_span
+  end in
+  let label_to_location = Int.Table.create () in
+  let location_to_label = Hashtbl.create (module LocSp) in
```

Comment:
@seantalts , this is the use of local state that I am talking about. How bad is this? Should I rewrite it using a purely functional traverse to enforce local data flow?

Reply:
Seems fine given that you return an immutable map at the end. Code is reasonably easy for me to read.

Reply:
Great! Thanks for looking at it. I just don't want to be religious about the purely functional style. I'd rather be pragmatic personally. Though I guess a point could be made for being consistent.